File: src/script/builder/ir/ir.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "./ir.h"

#include "../builder.h"

namespace tvm {
namespace script {
namespace builder {
namespace ir {

String GetUniqueName(const String& name) {
  // TODO: use tvm/ir/global_var_supply.h in PR #12066
  return name;
}

IRModuleFrame IRModule() {
  ObjectPtr<IRModuleFrameNode> n = make_object<IRModuleFrameNode>();
  n->functions.clear();
  return IRModuleFrame(n);
}

void IRModuleFrameNode::ExitWithScope() {
  Map<GlobalVar, BaseFunc> func_map;
  for (const auto& kv : functions) {
    const GlobalVar& gv = kv.first;
    const Optional<BaseFunc>& func = kv.second;
    CHECK(func.defined()) << "ValueError: function " << gv->name_hint << " is not defined";
    func_map.Set(gv, func.value());
  }
  Builder builder = Builder::Current();
  ICHECK(!builder->result.defined()) << "ValueError: Builder.result has already been set";
  builder->result = tvm::IRModule(func_map);
}

IRModuleFrame FindModuleFrame(const String& method) {
  Builder builder = Builder::Current();
  if (Optional<IRModuleFrame> frame = builder->FindFrame<IRModuleFrame>()) {
    if (builder->GetLastFrame<IRModuleFrame>().value() == frame) {
      return frame.value();
    }
  } else {
    LOG(FATAL) << "ValueError: IRModule frame not find. Please ensure '" << method
               << "' is called under I.ir_module()";
  }
  LOG(FATAL) << "ValueError: '" << method << "' must be called immediately under I.ir_module()";
  throw;
}

GlobalVar AddFunction(const String& func_name,         //
                      const Optional<BaseFunc>& func,  //
                      bool allow_rename) {
  IRModuleFrame frame = FindModuleFrame("I.AddFunction");
  if (!allow_rename) {
    CHECK(!frame->global_var_map.count(func_name))
        << "ValueError: function " << func_name << " already exists";
  }
  String name = GetUniqueName(func_name);
  GlobalVar gv = GlobalVar(name);
  frame->global_var_map.Set(name, gv);
  frame->functions.Set(gv, func);
  return gv;
}

void UpdateFunction(const String& func_name, const BaseFunc& func, bool require_first_define) {
  IRModuleFrame frame = FindModuleFrame("I.UpdateFunction");
  auto it = frame->global_var_map.find(func_name);
  CHECK(it != frame->global_var_map.end())
      << "ValueError: function " << func_name << " does not exist";
  const GlobalVar& gv = (*it).second;
  if (require_first_define) {
    CHECK(!frame->functions.at(gv).defined())
        << "ValueError: function " << func_name << " has already been defined";
  }
  frame->functions.Set(gv, func);
}

TVM_REGISTER_NODE_TYPE(IRModuleFrameNode);
TVM_REGISTER_GLOBAL("script.builder.ir.IRModule").set_body_typed(IRModule);
TVM_REGISTER_GLOBAL("script.builder.ir.AddFunction").set_body_typed(AddFunction);
TVM_REGISTER_GLOBAL("script.builder.ir.UpdateFunction").set_body_typed(UpdateFunction);

}  // namespace ir
}  // namespace builder
}  // namespace script
}  // namespace tvm

File: python/tvm/script/builder/ir/ir.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""TVMScript IR"""

import inspect
from typing import Optional, Type, Union

from tvm._ffi import register_object as _register_object
from tvm.ir import IRModule, BaseFunc

from ..frame import Frame
from . import _ffi_api


def is_defined_in_module(frames):
    if len(frames) > 2:
        maybe_class_frame = frames[2]
        statement_list = maybe_class_frame[4]
        first_statement = statement_list[0]
        line = first_statement.strip()
        if line.startswith("class "):
            return True
        if line.startswith("@") and "ir_module" in line:
            return True
    return False


@_register_object("script.builder.ir.IRModuleFrame")
class IRModuleFrame(Frame):
    ...


def ir_module(f: Optional[Type] = None) -> Union[IRModuleFrame, IRModule]:
    if f is not None:
        # pylint: disable=import-outside-toplevel
        from tvm.script.parse import parse
        from tvm.script.parse.utils import inspect_class_capture

        # pylint: enable=import-outside-toplevel

        if not inspect.isclass(f):
            raise TypeError(f"Expect a class, but got: {f}")

        return parse(f, inspect_class_capture(f))
    return _ffi_api.IRModule()  # pylint: disable=no-member # type: ignore


setattr(ir_module, "dispatch_token", "ir")


def add_function(name: str, func: Optional[BaseFunc] = None, allow_rename: bool = False) -> None:
    return _ffi_api.AddFunction(
        name, func, allow_rename
    )  # pylint: disable=no-member # type: ignore


def update_function(name: str, func: BaseFunc, require_first_define: bool=True) -> None:
    return _ffi_api.UpdateFunction(name, func, require_first_define)  # pylint: disable=no-member # type: ignore
